larpandoracontent/LArThreeDReco/LArTrackFragments/ClearTrackFragmentsTool.cc

```diff
@@ -52,52 +52,52 @@ bool ClearTrackFragmentsTool::FindTrackFragments(ThreeDTrackFragmentsAlgorithm *
+        const HitType fragmentHitType(overlapResult.GetFragmentHitType());
+        const Cluster *pClusterU(iter->GetClusterU()), *pClusterV(iter->GetClusterV()), *pClusterW(iter->GetClusterW());
 
-            const HitType fragmentHitType(overlapResult.GetFragmentHitType());
-            const Cluster *const pClusterU((TPC_VIEW_U == fragmentHitType) ? pFragmentCluster : (*iIter)->GetClusterU());
-            const Cluster *const pClusterV((TPC_VIEW_V == fragmentHitType) ? pFragmentCluster : (*iIter)->GetClusterV());
-            const Cluster *const pClusterW((TPC_VIEW_W == fragmentHitType) ? pFragmentCluster : (*iIter)->GetClusterW());
+        if (!this->CheckOverlapResult(overlapResult))
+            continue;
 
-            if (!(pClusterU->IsAvailable() && pClusterV->IsAvailable() && pClusterW->IsAvailable()))
-                throw StatusCodeException(STATUS_CODE_FAILURE);
+        // ATTN No longer guaranteed safe to use iterator after processing tensor element, hence caching results above
+        const Cluster *pFragmentCluster(nullptr);
+        this->ProcessTensorElement(pAlgorithm, overlapTensor, overlapResult, pFragmentCluster);
 
-            clustersToRemoveFromTensor.push_back(pClusterU);
-            clustersToRemoveFromTensor.push_back(pClusterV);
-            clustersToRemoveFromTensor.push_back(pClusterW);
-
-            ProtoParticle protoParticle;
-            ProtoParticleVector protoParticleVector;
-            protoParticle.m_clusterListU.push_back(pClusterU);
-            protoParticle.m_clusterListV.push_back(pClusterV);
-            protoParticle.m_clusterListW.push_back(pClusterW);
-            protoParticleVector.push_back(protoParticle);
-            particlesMade |= pAlgorithm->CreateThreeDParticles(protoParticleVector);
-        }
+        if (!pFragmentCluster)
+            throw StatusCodeException(STATUS_CODE_FAILURE);
 
-        clustersToRemoveFromTensor.insert(clustersToRemoveFromTensor.end(), modifiedClusters.begin(), modifiedClusters.end());
-        clustersToRemoveFromTensor.insert(clustersToRemoveFromTensor.end(), deletedClusters.begin(), deletedClusters.end());
+        if (TPC_VIEW_U == fragmentHitType) pClusterU = pFragmentCluster;
+        if (TPC_VIEW_V == fragmentHitType) pClusterV = pFragmentCluster;
+        if (TPC_VIEW_W == fragmentHitType) pClusterW = pFragmentCluster;
 
-        this->RebuildClusters(pAlgorithm, modifiedClusters, clustersToAddToTensor);
-        this->UpdateTensor(pAlgorithm, overlapTensor, clustersToRemoveFromTensor, clustersToAddToTensor);
+        if (!(pClusterU->IsAvailable() && pClusterV->IsAvailable() && pClusterW->IsAvailable()))
+            throw StatusCodeException(STATUS_CODE_FAILURE);
 
-        if (particlesMade)
-            return true;
+        // ATTN For safety, remove all clusters associated with this fragment particle from the tensor
+        ClusterList fragmentClusterList, affectedKeyClusters;
```

Comment:
This is a minor comment, but what is the difference between clusters and keyClusters?  I believe keyClusters are a subset of the fragmented clusters that have sufficient length and number of hits to be considered in the overlap tensor.  Is this correct?

Reply:
Thanks a lot for looking! Yes, the keyClusters are those clusters upon which the tensor is keyed. The tensor is like a big, nested map, keyed on U clusters, then V clusters, then W clusters and finally having an OverlapResult as its ultimate value.

Most algorithms will place some quality cuts on the clusters that are used as key clusters. Typically, the key clusters are the only clusters that are ever considered, and which will be modified and/or used to build particles.

In this one, ambitious alg/tool, however, the plan was to link good key clusters in two views, e.g. V, W and associate them to lots of fragment clusters in the third view e.g. U.

It's still a good idea, and it makes good changes (well, examining event displays during development suggested this), but you can see how this makes things more complicated, especially as the alg/tool together request the fragments to be reclustered and merged, and then needs to update the tensor for the next passes!